fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/RecordTypesProperty.java

```diff
@@ -57,115 +57,136 @@
-                    }
-                    return union;
-                } else {
-                    throw new RecordCoreException("tried to find record types for a relational expression with multiple " +
-                                                  "relational children, but no combiner was specified");
-                }
-            }
-        }
+    public RecordTypesVisitor createVisitor() {
+        return new RecordTypesVisitor(Optional.empty());
     }
 
-    @Nonnull
     @Override
-    public Set<String> evaluateAtRef(@Nonnull Reference ref,
-                                     @Nonnull List<Set<String>> memberResults) {
-        final Set<String> union = new HashSet<>();
-        for (Set<String> resultSet : memberResults) {
-            union.addAll(resultSet);
-        }
-        return union;
+    public String toString() {
+        return getClass().getSimpleName();
     }
 
     @Nonnull
-    public static Set<String> evaluate(@Nonnull Reference ref) {
-        return Objects.requireNonNull(ref.acceptPropertyVisitor(new RecordTypesProperty(Optional.empty())));
+    public Set<String> evaluate(@Nonnull Reference reference) {
+        return Objects.requireNonNull(reference.acceptPropertyVisitor(createVisitor()));
     }
 
     @Nonnull
-    public static Set<String> evaluate(@Nonnull AliasResolver aliasResolver,
-                                       @Nonnull Reference ref) {
-        return Objects.requireNonNull(ref.acceptPropertyVisitor(new RecordTypesProperty(Optional.of(aliasResolver))));
+    public Set<String> evaluate(@Nonnull final RelationalExpression expression) {
+        return Objects.requireNonNull(expression.acceptPropertyVisitor(createVisitor()));
     }
 
     @Nonnull
-    public static Set<String> evaluate(@Nonnull AliasResolver aliasResolver,
-                                       @Nonnull RelationalExpression ref) {
-        return Objects.requireNonNull(ref.acceptPropertyVisitor(new RecordTypesProperty(Optional.of(aliasResolver))));
+    public static RecordTypesProperty recordTypes() {
+        return RECORD_TYPES;
+    }
+
+    public static class RecordTypesVisitor implements SimpleExpressionVisitor<Set<String>> {
+        @Nonnull
+        private final Optional<AliasResolver> aliasResolverOptional;
+
+        public RecordTypesVisitor(@Nonnull final AliasResolver aliasResolver) {
+            this(Optional.of(aliasResolver));
+        }
+
+        public RecordTypesVisitor(@Nonnull final Optional<AliasResolver> aliasResolverOptional) {
+            this.aliasResolverOptional = aliasResolverOptional;
+        }
+
+        @Nonnull
+        @Override
+        public Set<String> evaluateAtExpression(@Nonnull RelationalExpression expression, @Nonnull List<Set<String>> childResults) {
+            // shouldVisit() ensures that we only visit relational planner expressions
+            // If we mess this up, better to find out sooner rather than later.
```

Comment:
This comment was there in the old code, but I'm not sure I understand it. What does it mean by "relational" planner expression here? That is, does it think there's some kind of `RelationalExpression` that is not a "relational planner expression"?

Reply:
I can make this explanation better. In short, there is a kind of visitor (which does not really follow the strict visitor pattern and which predates the visitors that extend the code-generated visitors) that allows you to traverse the graph while skipping subgraphs. This is what the `shouldVisit()` indicates. If `shouldVisit()` is `true`, this method must return something useful.

Reply:
Changed my mind and just removed the comment.